### ISSUE TITLE
Memory leak in scan macros

### ISSUE THREAD

**Issue by jkotan** [sardana-org/sardana#1095]
Hi, 

I've just tested MacroServer with 1D channel detector (MCA) and after a less than one day running MacroServer uses 4.6g memory (It started to be proportional to number of point in my scan I've performed). 

To look at this problem in more details I create macros:
```python
@macro()
def getobjgraph(self):

    import gc
    import objgraph
    mystdout = StringIO()
    objgraph.show_most_common_types(file=mystdout)
    self.output(str(mystdout.getvalue()))


@macro()
def gcdump(self):

    import gc
    gc.set_debug(gc.DEBUG_LEAK)
    gc.collect()
    gres = gc.garbage
    self.output("GARBAGE %s" % len(gres))
    res = str(gres)
    import time
    fname = "/tmp/gcdump-%s.gc" % str(time.time())
    self.output(fname)
    fl = open(fname, "w")
    fl.write(res)
    fl.close()
```  
from `gcdump` where I see that a number of garbages grows proportionally to a number of scan points I've performed (with proportional factor: 3)

On the other hand from `getobjgraph`  I see that .e.g. 
```
p09/door/haso228jk.01 [36]: %getobjgraph
dict               86524
CaselessDict       71883
Record             71828
function           27811
list               15452
tuple              12674
weakref            7533
wrapper_descriptor 5856
getset_descriptor  4476
type               3116
```
where `Record`, `CaselessDict` and `dict` are amost equal (up to a constant factor) to number of scan points I've performed with `ascan` macro. 

Any idea what it comes from? Something is not deleted? (I use sardana 2.5.1)

**Comment by reszelaz** [comment]
Thanks for the report @jkotan !

I have not yet tested your macros - I will certainly try them! But I have already some questions.
In your test do you run a single long scan? During the scan all the records are kept in the memory. Furthermore, by default, they are even kept in the memory after the scan finish in order to allow the scan data analysis in the subsequent macro. This second feature can be however disabled with [PreserveMacroData](https://sardana-controls.org/users/environment_variable_catalog.html#preservemacrodata) environment variable.

If this still does explain this issue, for example you run a sequence of short scans and the memory sill grows, this may be related to some cyclic references that makes the life harder to the garbage collector. #816 tries to fix them. This PR was already approved but before merging it I wanted to carefully review the public API that this PR touches to not break anything. It should be available soon.

**Comment by jkotan** [comment]
Thanks @reszelaz ,
In the above case I ran 7 scans with about 10000 points per each.The memory usage was  growing with (# scan) proportionally.

Indeed, in order to limit memory leak one has to do both: set `PreserveMacroData=False` and use #816. With them usage of the memory grows very slow and numbers of object in gc.garbage is no more proportional too number of scan points

**Comment by guifrecuni** [comment]
Thanks @jkotan , it puzzles me that the usage of the PreserveMacroData env variable produces a memory leak proportional to the number of scans... we will have to investigate it.

**Comment by reszelaz** [comment]
@jkotan #816 is already merged into develop.

**Comment by jkotan** [comment]
Thanks

**Comment by jkotan** [comment]
@guifrecuni, with `PreserveMacroData=False` usage of memory by MacroServer grows with the number of points and scans but it looks like it is not a memory leak, i.e. GC does not show this memory as uncollectable. 
For me this memory usage for the above scan is on the level ~ 33kB per point. So after to scans with 10000 points `MacroServer` uses ~ 800MB (about 330MB per each scan).
Is it natural for `MacroServer`?

**Comment by jkotan** [comment]
The above 33kB  memory usage corresponds to size of the data record (one mca with 4096 channels of double type + counter). If I add second mca  the usage of memory grows by 32kB per each step (to 65kB per step)

Moreover, when I switch from sardana 2.5.1 to sardana 2.7.2 (to test #816) MacroServer hung a few time without a reason wtih
```
Thread 43 "python" received signal SIGPIPE, Broken pipe.
[Switching to Thread 0x7fff47fff700 (LWP 9608)]
0x00007ffff7bcc63f in __libc_send (fd=36, buf=0x7fff58002e98, n=186, flags=0)
    at ../sysdeps/unix/sysv/linux/x86_64/send.c:26
26	../sysdeps/unix/sysv/linux/x86_64/send.c: No such file or directory.
(gdb) bt
#0  0x00007ffff7bcc63f in __libc_send (fd=36, buf=0x7fff58002e98, n=186, flags=0) at ../sysdeps/unix/sysv/linux/x86_64/send.c:26
#1  0x00007ffff320300b in omni::unixConnection::Send(void*, unsigned long, unsigned long, unsigned long) () from /usr/lib/libomniORB4.so.1
#2  0x00007ffff31c5344 in omni::giopStream::sendChunk(omni::giopStream_Buffer*) () from /usr/lib/libomniORB4.so.1
#3  0x00007ffff31d8283 in omni::giopImpl12::outputMessageEnd(omni::giopStream*) () from /usr/lib/libomniORB4.so.1
#4  0x00007ffff31dc918 in omni::giopImpl12::sendUserException(omni::giopStream*, CORBA::UserException const&) () from /usr/lib/libomniORB4.so.1
#5  0x00007ffff31cc21b in omni::GIOP_S::handleRequest() () from /usr/lib/libomniORB4.so.1
#6  0x00007ffff31cca98 in omni::GIOP_S::dispatcher() () from /usr/lib/libomniORB4.so.1
#7  0x00007ffff31c96c5 in omni::giopWorker::real_execute() () from /usr/lib/libomniORB4.so.1
#8  0x00007ffff31c9d4f in omni::giopWorker::execute() () from /usr/lib/libomniORB4.so.1
#9  0x00007ffff318002d in omniAsyncWorkerInfo::run() () from /usr/lib/libomniORB4.so.1
#10 0x00007ffff4abfc4c in Tango::create_PyPerThData (info=...) at utils.cpp:3234
#11 0x00007ffff317ff24 in omniAsyncWorkerInfo::run() () from /usr/lib/libomniORB4.so.1
#12 0x00007ffff31805ff in omniAsyncWorker::run(void*) () from /usr/lib/libomniORB4.so.1
#13 0x00007ffff2ec16d9 in omni_thread_wrapper () from /usr/lib/libomnithread.so.3
#14 0x00007ffff7bc34a4 in start_thread (arg=0x7fff47fff700) at pthread_create.c:456
#15 0x00007ffff6fe0d0f in clone () at ../sysdeps/unix/sysv/linux/x86_64/clone.S:97
```
(which I haven't noticed with 2.5.1). Maybe it is related to #1023

**Comment by jkotan** [comment]
I've performed a few test tracking where the huge memory usage comes from and  I found that it is caused by  https://github.com/sardana-org/sardana/blob/d6b68281b3813c1232fcac92efc4ed6cfbd479c1/src/sardana/macroserver/scan/scandata.py#L390
However, the `records` list is cleaned by python so that is not a memory leak.
Thus:
1) If I remove `append` to `records` and a few following command there is no more huge memory usage
2) but if I  remove `append` to `records` but start to append `records` to a new global list  `records2`,  (which I clean after each scan) I've got again  the huge memory usage related to `records2`
3) However,  if I start to append very big arrays to `records2`  the huge memory usage disappear (I see only huge usage of virtual memory).

Finally, I found 
https://www.reddit.com/r/Python/comments/5sf6h5/python_does_not_return_memory_to_linux/
which explains the problem, i.e. in complicated programs it looks like that it is impossible for `cpython` (because of `malloc()` limitations) to return memory to the system which was reservered in small blocks. 

Currently to avoid this problem one could either reserve memory for scan records in a big block or use storage a separete process. But both solutions are hard to implement in python.

**Comment by guifrecuni** [comment]
Good catch @jkotan , thanks for the reference.
I propose to discuss about it in one of our follow-up meetings and take a decision on how we should proceed.

**Comment by jkotan** [comment]
I also found that starting MacroServer with `jemalloc`, i.e.
```
LD_PRELOAD=/usr/lib/x86_64-linux-gnu/libjemalloc.so.1 MacroServer instance
```
the usage of memory is much lower 

https://zapier.com/engineering/celery-python-jemalloc/

**Comment by reszelaz** [comment]
Hi Jan,

First many thanks for your investigation on this problem! We saw and analyzed it long time ago coming to similar conclusions. Now we revisit this problem at ALBA since we run continues scans with many points >10k and some with 1D (spectrum) channels.

I confirm what you say about the `RecordList` object. In the case of the continues scan, the memory increases only if we use this class. When I commented its usage and avoid usege of any type of recorder the memory stays stable.

I also made tests with the `jemalloc` which comes with Debian stretch. I use `timescan` of 10k points, 1 ms integration time and 11 dummy C/T (with simulated hardware synchronization). 

Here are the results with the standard libc implementation:

![ms_mem_10rep_10k_11ch_manual](https://user-images.githubusercontent.com/6735649/76153360-31e5e900-60cb-11ea-90f8-2b2da6fce2e5.png)

And the same test with jemalloc:

![ms_jemalloc_mem_10rep_10k_11ch_manual](https://user-images.githubusercontent.com/6735649/76153366-40340500-60cb-11ea-8b43-1df31aa7199d.png)

Note that when I run very fast repetitions of the same scans (the above executions were done manually from spock with some delay in between scans), the behavior in the case of libc implementation is different:

![ms_mem_10rep_10k_11ch_repeat](https://user-images.githubusercontent.com/6735649/76153379-6f4a7680-60cb-11ea-9b75-8070253bb980.png)

While the jemalloc does not change:

![ms_jemalloc_mem_10rep_10k_11ch_repeat](https://user-images.githubusercontent.com/6735649/76153381-825d4680-60cb-11ea-9f1f-543c65952443.png)

Finally, do you use jemalloc at DESY? Which version? What is your experience with that?

Many thanks!

**Comment by jkotan** [comment]
Hi @Zibi,
With which version of sardana/Taurus you have performed the tests ?

**Comment by reszelaz** [comment]
The latest develop branches of both.

**Comment by reszelaz** [comment]
> When I commented its usage and avoid usege of any type of recorder the memory stays stable.

I just noticed that I made these test with value buffer events disabled. These are by default serialized/deserialized with pickle. And I read that this may also require some extra memory. I will check it back on Monday.

**Comment by jkotan** [comment]
I've made my tests with python 2.7. In DESY with still use py2 with standard jemalloc from stretch which decreases memory usage. However, py3 has improved GC e.g. removes circular dependent objects. So it may happen that for py3 we don't need jemalloc anymore. I will check it next week

**Comment by reszelaz** [comment]
> I just noticed that I made these test with value buffer events disabled. These are by default serialized/deserialized with pickle. And I read that this may also require some extra memory.

I repeated this test. unpickle in the value buffer event handling does not affect the memory of the MacroServer. Note it still may affect the Pool where we pickle. But let's focus for the moment on the MacroServer.

I think all the memory that MacroServer can not free comes from the `RecordList` handling (`dict`). In the case of the continuous scans this also includes generation of theoretical timestamps and filling of missing records but at the end all this is for having the complete `RecordList`. I don't think it is related with the GC and circular references but as your links point to the way the memory is allocated `brk` vs. `mmap`. That's why using jemalloc, which uses the second one, we can see the improvement. And it looks like in Python3 this is the same as in Python2. Tomorrow we will try using jemalloc at the beamline since you have a possitive experience with it at DESY. We will keep you updated!

**Comment by tiagocoutinho** [comment]
> Currently to avoid this problem one could either reserve memory for scan records in a big block or use storage a separete process. But both solutions are hard to implement in python.

For scans with known number of points we could reserve memory in one (or several) numpy array(s). I guess that since this is pure C allocation without direct intervention from the python memory management system the problem would be solved. I don't think it would be too difficult to implement, do you?

For timescan memory allocation could be done by blocks and at the end use [numpy.concatenate](http://docs.scipy.org/doc/numpy/reference/generated/numpy.concatenate.html)/[numpy.column_stack](http://docs.scipy.org/doc/numpy/reference/generated/numpy.column_stack.html) at the end.

**Comment by reszelaz** [comment]
Hi Jan,
We were running for a week the Pool and the MS of one of our beamlines with the jemalloc and the mem leak is much slower. We plan to apply it for other beamlines. One question, do you use jemalloc also for other PyTango DS?
Thanks!

**Comment by jkotan** [comment]
We have only switched to jemalloc for Sardana DSs and our NeXus DSs.

**Comment by tiagocoutinho** [comment]
Going back on my previous comment just a detail: probably since scans may have columns that reference 1D/2D data location, numpy arrays are not sufficient. We may need to use [pandas.DataFrame](https://pandas.pydata.org/docs/reference/frame.html). @cpascual, you have experience with pandas and also sardana recorders: would you agree?

**Comment by felix92** [comment]
Hi all,

we have noticed the same problem with our 2d camera controllers, that after the scans the memory is not freed again. If the scans are repeated, the memory usage increases until the memory is completely full.
We tried also scans with the BasicDummyTwoDController with the same results. The data is saved in a hdf5 file.

We tried to fix it with `PreserveMacroData=False` and also with `jamalloc` but this had no influnce on this problem.

We also tried it on two different systems to check if different versions have an influence but this problem occurred on both. Here are the versions.

```
Linux faraday 5.0.0-32-generic #34~18.04.2-Ubuntu SMP Thu Oct 10 10:36:02 UTC 2019 x86_64 x86_64 x86_64 GNU/Linux
Python 3.6.9
Sardana 3.0.3

```
```
Linux tesla 5.4.0-74-generic #83-Ubuntu SMP Sat May 8 02:35:39 UTC 2021 x86_64 x86_64 x86_64 GNU/Linux
Python 3.8.10
Sardana 3.1.1-alpha
```

Maybe one of you could try to run several scans with 100 data points with the BasicDummyTwoDController and can check if the memory is freed after each scan. 

Thanks,
Felix

**Comment by reszelaz** [comment]
Hi @felix92,

Thanks for your feedback. Very recently @saidaHF started investigating this issue. She focuses the investigation on: https://github.com/sardana-org/sardana/blob/0307007258d34055606d6e0fffccda42e7f1f33b/src/sardana/macroserver/scan/scandata.py#L296
I believe it should possible to reproduce the issue with a much simpler python program e.g. https://github.com/reszelaz/storage-kpi/blob/py3/sardana_scan.py.
Currently, this script only simulates scans with scalar data but we will also try it with images and will get back to you soon.

As @tiagocoutinho proposes in https://github.com/sardana-org/sardana/issues/1095#issuecomment-608306064 we will try to substitute this class with the `pandas.DataFrame` and see if the memory is released correctly.

Cheers!

**Comment by dschick** [comment]
Hi @reszelaz ,

thanks for your promt reply.
We are actually wondering that the `jemalloc` fix is not working for us.
Is this maybe due to the fact that we are using 2D channles vs. the 1D channles discussed above?
We tested with jemalloc 5.1 from the Mint repos.

I quickly test your script 

>  https://github.com/reszelaz/storage-kpi/blob/py3/sardana_scan.py.

and introduced some changes:
```python
def _prepare_environ(self):
        self.environ = dict()
        self.environ["startts"] = ts = time.time()
        self.environ["starttime"] = datetime.datetime.fromtimestamp(ts)
        self.environ["serialno"] = 0
        data_desc = list()
        data_desc.append(ColumnDesc(name="point_nb",
                                    label="#Pt No",
                                    dtype="int64"))
        for i in range(self.nb_of_columns):
            data_desc.append(ColumnDesc(name="col%d" % i,
                                        label="col%d" % i,
                                        dtype="float64"))
        data_desc.append(ColumnDesc(name="image",
                                    label="image",
                                    dtype="(float64,)"))
        data_desc.append(ColumnDesc(name="timestamp",
                                    label="dt",
                                    dtype="float64"))
        self.environ["datadesc"] = data_desc
        self.environ["title"] = DESCRIPTION
        self.environ["user"] = "sicilia"
```
and
```python
def _acquire(self, point_nb):
        time.sleep(self.integ_time)
        data_line = dict()
        data_line["point_nb"] = point_nb
        for col in range(self.nb_of_columns):
            value = random.random() * random.randint(1, 10)
            data_line["col%d" % col] = value
        timestamp = point_nb * random.random()
        data_line["image"] = np.ones((2048, 2048))*32000
        data_line["timestamp"] = timestamp
        return data_line
```
and in the `main()` function
```python
def main():
    parser = argparse.ArgumentParser(description=DESCRIPTION)
    parser.add_argument("-f, ", "--file", metavar="file", type=str, nargs="?",
                        required=True, help="File to store the scan data")
    args = parser.parse_args()
    file_ = args.file
    scan = Scan(file_)
    for i in range(10):
        scan.start()
        scan.run()
        scan.end()
```

This is actually working pretty well and the memory is always freed at the end of ever Scan.
So `self.record_list.end()` seems to work correctly here.

**Comment by reszelaz** [comment]
I will check your discoveries with the script and 2D on Monday. I wonder if we have introduced a regression in this term recently. On our beamlines at ALBA we recently suffer much worse memory consumption performance of the MacroServer. I think it was not so bad using Sardana from the first months of 2020 (already after migration to Python 3, but I think still before the official 3.0.3 release). Could you for example check if using this revision: 391c1102377dec639241e604c7b602a2813cf4f0 the memory increase is the same?

**Comment by saidaHF** [comment]
Hi @felix92, @dschick,

I created a measurement group: `defmeas mntgrp_memleak_2d twod01`
Then, I selected it: `set_meas mntgrp_memleak_2d`

I am using this macro:

`repeat 10 "ascan mot01 0 100 100 0.001"`

And I can see this result in my taurus trend using a 2D channel:


In this case, [PreserveMacroData](https://sardana-controls.org/users/environment_variable_catalog.html#preservemacrodata) is **True**

![image](https://user-images.githubusercontent.com/73069754/126457296-8c391750-906e-4854-869e-c24ecca7a317.png)

This scan goes up to 2GB approximately

Then, I did the same scan but when **PreserveMacroData** is **False**

`senv PreserveMacroData False`

And goes up to 1GB approximately, less than before as we can see:

![image](https://user-images.githubusercontent.com/73069754/126457343-be7f0084-e6b1-41e8-bf96-0d0e42c8ef54.png)

I think this is a normal memory usage, because only 2D data needs more than 800 MB

we can see this memory usage in spock:

```python
import sys


def get_scan_data_size(scan_data):
    size = 0
    for pt_nb, record in scan_data.items():
        for name, value in record.data.items():
            try:
                size += value.nbytes
            except AttributeError:
                size += sys.getsizeof(value)
    return size


# here we need to shorthen the scan to just 10 nb_interv
ascan mot01 0 10 10 0.1

# and increase the CORBA timeout to 60 s instead of 3 s
_DOOR.set_timeout_millis(60000)

data = %macrodata

# It uses 92 MB in a 10 interval scan. So the 100 interval scan would have 920 MB
print(get_scan_data_size(data))
```

So, I have not been able to observe a constant increase in memory.

**Comment by dschick** [comment]
Hi @saidaHF ,

many thanks for testing.
Have you also enabled saving the data to file?
If I remember correctly this issue was related to the actual records in the Recorder.

The setting of the `PreserveMacroData` also worked well for us.

BTW: how do you plot the memory usage of the process via Taurus?

Best

Daniel

**Comment by saidaHF** [comment]
# **Saving the data to file**

Hi @dschick,

I use this command to open taurus trend and monitor the memory usage:

`taurus trend "eval:@psutil.*/Process(<MacroServer PID>).memory_info().rss"`

( You can see your PID with `ps -ef | grep <MacroServer name> `).

Then I repeated the process to see the file's problem

`repeat 10 "ascan mot01 0 100 100 0.001"` 

PreserveMacroData = True

![](https://i.imgur.com/6rFXBcu.png)

I saved in **/tmp/scanfiletmp2 (Spec)** and I can see the data saved in this file with 
`less /tmp/scanfiletmp2` in today's date (07/21/2021)

![](https://i.imgur.com/74PjDrz.png)

And I repeat the same scan with PreserveMacroData = False

![](https://i.imgur.com/zWniHPU.png)

It also saves the data in **/tmp/scanfiletmp2 (Spec)**

![](https://i.imgur.com/5jXwUwH.png)

May I know what recorder are you using? 

Thanks!

**Comment by dschick** [comment]
thanks for the snippet and the test with the SPEC recorder.
I am actually using the h5 recorder. My results look like
![grafik](https://user-images.githubusercontent.com/13797420/126493833-4aec707c-5c47-4b03-834f-88099dfddb02.png)

I ran the same macro as you.

**Comment by saidaHF** [comment]
Hi @felix92 @dschick 

I used a measurement group with these channels: 

`defmeas mesgroup ct01 ct02 ct03 ct04 ct05 ct06 ct07 ct08 ct09 ct10 oned01 twod01`

with the same scans as before:

`repeat 10 "ascan mot01 0 100 100 0.001"`

And I saved this information in two files:  **/tmp/scans6.h5 (HDF5::NXscan)** and **/tmp/scans6.dat (Spec)**

We can show this result:

![image](https://user-images.githubusercontent.com/73069754/127316279-bca9baf8-0502-4c5d-b981-d4f4a186c738.png)

where the memory stays constant

**Comment by dschick** [comment]
many thanks @saidaHF 

I justed tested your fix and it seems to work.
However, in my case the memory seems to be freed only after every second scan, and stays constant otherwise.

![grafik](https://user-images.githubusercontent.com/13797420/127390108-129ed7f4-945d-48f1-86fa-32cc09c6e3d9.png)

are you using jemalloc for your tests?

**Comment by reszelaz** [comment]
> However, in my case the memory seems to be freed only after every second scan, and stays constant otherwise.

Hi @dschick,
Many thanks for testing this PR! It looks like @saidaHF run her tests with `PreserveMacroData=False`.

**Comment by dschick** [comment]
okay, then I think this issue is solved and the PR can be merged!

Thanks a lot!

**Comment by reszelaz** [comment]
#1669 that fixes the _mem leak_ with the `NXscanH5_FileRecorder` is already merged.
Soon #1664 will get merged solving the _mem leak_ with the continuous scans (ascanct, etc.). As soon as this is done I propose to close this issue. The memory of the MacroServer won't grow beyond what is needed for the scan data (there is no progressive accumulation) and IMHO it is not urgent now to look for other solutions here.

**Comment by felix92** [comment]
Hello @saidaHF,
can you try to repeat `repeat 10 "ascan mot01 0 100 100 0.001"` three or more times? We observed that the memory will increase after each repeat of this macro. The same happens if you run several times `ascan mot01 0 100 100 0.001` e.g. in a for loop.

Tomorrow I can send some memory plots for different situations.

Best,
Felix

**Comment by saidaHF** [comment]
Hi @felix92 

I repeat three times this loop:

```
for i in range(10): 
    %ascan mot01 0 100 100 0.001
```
This is like 30 ascans, and the same thing happens to me:

![image](https://user-images.githubusercontent.com/73069754/127634261-c0929e0b-bd09-4c38-b42b-d4349738fe25.png)

Thanks.

**Comment by felix92** [comment]
Hi,
thanks for testing.
What still confuses me a bit is when you start a scan the increase  at the beginning is very strong and then flattens. Actually, the memory consumption should be the same for each measuring point and therefore the large increase at the beginning doesn't make sense to me.
Secondly, it is noticeable that at the end of the scan the memory does not return to the initial situation. If the memory is released again after the scan, then the trend should be back to the level before the scan.

Do I see this correctly?

Best,
Felix

**Comment by reszelaz** [comment]
Thanks @saidaHF, @felix92 and @dschick for your recent tests!
I think we are getting closer to the root cause of the problem.
I have done some tests without running macros, and just using the `RecordList` class in a simulated scan as commented in https://github.com/sardana-org/sardana/issues/1095#issuecomment-875899041.
I have enhanced this script with the @dschick suggestions to use 2D data from https://github.com/sardana-org/sardana/issues/1095#issuecomment-875968305.
And I have discovered that the problems appear when we start using [`taurus.core.util.threadpool.ThreadPool`](https://gitlab.com/taurus-org/taurus/-/blob/develop/lib/taurus/core/util/threadpool.py#L42). And this is how macros are executed withing MacroServer.
When I run the simulated scan using `concurrent.futures.ThreadPoolExecutor` or a single `threading.Thread` the memory never exceeds the amount necessary for one scan. And when we do use threads at all, the memory is released completelly:

```
┬─[zreszela@pt222:~/w/s/src]─[07:55:21 PM]─[G:py3↑7|✚1]
╰─>$ python sardana_scan.py -f /tmp/scan.dat -r 10 -t taurus                                                                                                                       (sardana) 
repeat #1
RSS before scan: 78.95 MB
RSS after scan: 81.73 MB
repeat #2
RSS before scan: 81.73 MB
RSS after scan: 865.9 MB
repeat #3
RSS before scan: 865.9 MB
RSS after scan: 1.61 GB
repeat #4
RSS before scan: 1.61 GB
RSS after scan: 2.38 GB
repeat #5
RSS before scan: 2.38 GB
RSS after scan: 3.14 GB
repeat #6
RSS before scan: 3.14 GB
RSS after scan: 3.91 GB
repeat #7
RSS before scan: 3.91 GB
RSS after scan: 4.67 GB
repeat #8
RSS before scan: 4.67 GB
RSS after scan: 5.44 GB
repeat #9
RSS before scan: 5.44 GB
RSS after scan: 6.21 GB
repeat #10
RSS before scan: 6.21 GB
RSS after scan: 6.99 GB
RSS final: 6.99 GB
┬─[zreszela@pt222:~/w/s/src]─[07:59:35 PM]─[G:py3↑7|✚1]
╰─>$ python sardana_scan.py -f /tmp/scan.dat -r 10 -t concurrent                                                                                                                   (sardana) 
repeat #1
RSS before scan: 78.93 MB
RSS after scan: 79.71 MB
repeat #2
RSS before scan: 79.71 MB
RSS after scan: 359.73 MB
repeat #3
RSS before scan: 359.73 MB
RSS after scan: 807.76 MB
repeat #4
RSS before scan: 807.76 MB
RSS after scan: 807.76 MB
repeat #5
RSS before scan: 807.76 MB
RSS after scan: 807.76 MB
repeat #6
RSS before scan: 807.76 MB
RSS after scan: 807.76 MB
repeat #7
RSS before scan: 807.76 MB
RSS after scan: 807.76 MB
repeat #8
RSS before scan: 807.76 MB
RSS after scan: 807.76 MB
repeat #9
RSS before scan: 807.76 MB
RSS after scan: 807.76 MB
repeat #10
RSS before scan: 807.76 MB
RSS after scan: 807.76 MB
RSS final: 807.76 MB
┬─[zreszela@pt222:~/w/s/src]─[07:59:44 PM]─[G:py3↑7|✚1]
╰─>$ python sardana_scan.py -f /tmp/scan.dat -r 10 -t thread                                                                                                                       (sardana) 

repeat #1
RSS before scan: 79.16 MB
RSS after scan: 79.95 MB
repeat #2
RSS before scan: 79.95 MB
RSS after scan: 527.97 MB
repeat #3
RSS before scan: 527.97 MB
RSS after scan: 808.0 MB
repeat #4
RSS before scan: 808.0 MB
RSS after scan: 808.0 MB
repeat #5
RSS before scan: 808.0 MB
RSS after scan: 808.0 MB
repeat #6
RSS before scan: 808.0 MB
RSS after scan: 808.0 MB
repeat #7
RSS before scan: 808.0 MB
RSS after scan: 808.0 MB
repeat #8
RSS before scan: 808.0 MB
RSS after scan: 808.0 MB
repeat #9
RSS before scan: 808.0 MB
RSS after scan: 808.0 MB
repeat #10
RSS before scan: 808.0 MB
RSS after scan: 808.0 MB
RSS final: 808.0 MB
┬─[zreszela@pt222:~/w/s/src]─[07:59:53 PM]─[G:py3↑7|✚1]
╰─>$                                                                                                                                                                               (sardana) 
┬─[zreszela@pt222:~/w/s/src]─[07:59:53 PM]─[G:py3↑7|✚1]
╰─>$ python sardana_scan.py -f /tmp/scan.dat -r 10 -t no                                                                                                                           (sardana) 
repeat #1
RSS before scan: 79.03 MB
RSS after scan: 79.82 MB
repeat #2
RSS before scan: 79.82 MB
RSS after scan: 79.92 MB
repeat #3
RSS before scan: 79.92 MB
RSS after scan: 79.92 MB
repeat #4
RSS before scan: 79.92 MB
RSS after scan: 79.92 MB
repeat #5
RSS before scan: 79.92 MB
RSS after scan: 79.92 MB
repeat #6
RSS before scan: 79.92 MB
RSS after scan: 79.92 MB
repeat #7
RSS before scan: 79.92 MB
RSS after scan: 79.92 MB
repeat #8
RSS before scan: 79.92 MB
RSS after scan: 79.92 MB
repeat #9
RSS before scan: 79.92 MB
RSS after scan: 79.92 MB
repeat #10
RSS before scan: 79.92 MB
RSS after scan: 79.92 MB
RSS final: 79.92 MB
```
I will continue next week...

**Comment by reszelaz** [comment]
Investigating further this issue I think that this memory leak is not due to the fact of using the taurus `ThreadPool` but threads in general.

In https://github.com/sardana-org/sardana/issues/1095#issuecomment-890069171 we could see that when using `concurrent.futures.ThreadPoolExecutor` the memory never exceeded the amount of memory necessary to store the scan data of one scan. The difference between the `concurrent.futures.ThreadPoolExecutor` and the taurus `ThreadPool` is that the first one uses the same thread as far as there is only one parallel job (see prints with `Scan run in <thread ID>`):
```
╰─>$ python sardana_scan.py -f /tmp/scan.dat -r 10 -t concurrent
repeat #1
RSS before scan: 104.05 MB
Scan run in 139865956955904
RSS after scan: 104.94 MB
repeat #2
RSS before scan: 104.94 MB
Scan run in 139865956955904
RSS after scan: 832.99 MB
repeat #3
RSS before scan: 832.99 MB
Scan run in 139865956955904
RSS after scan: 832.99 MB
repeat #4
RSS before scan: 832.99 MB
Scan run in 139865956955904
RSS after scan: 832.99 MB
repeat #5
RSS before scan: 832.99 MB
Scan run in 139865956955904
RSS after scan: 832.99 MB
repeat #6
RSS before scan: 832.99 MB
Scan run in 139865956955904
RSS after scan: 832.99 MB
repeat #7
RSS before scan: 832.99 MB
Scan run in 139865956955904
RSS after scan: 832.99 MB
repeat #8
RSS before scan: 832.99 MB
Scan run in 139865956955904
RSS after scan: 832.99 MB
repeat #9
RSS before scan: 832.99 MB
Scan run in 139865956955904
RSS after scan: 832.99 MB
repeat #10
RSS before scan: 832.99 MB
Scan run in 139865956955904
RSS after scan: 832.99 MB
RSS final: 832.99 MB
```
and taurus `ThreadPool` redistribute the jobs randomly:

```
╰─>$ python sardana_scan.py -f /tmp/scan.dat -r 10 -t taurus         
repeat #1
RSS before scan: 104.89 MB
Scan run in 140308756354816
RSS after scan: 107.96 MB
repeat #2
RSS before scan: 107.96 MB
Scan run in 140308747962112
RSS after scan: 892.13 MB
repeat #3
RSS before scan: 892.13 MB
Scan run in 140308739569408
RSS after scan: 1.64 GB
repeat #4
RSS before scan: 1.64 GB
Scan run in 140308528494336
RSS after scan: 2.4 GB
repeat #5
RSS before scan: 2.4 GB
Scan run in 140308520101632
RSS after scan: 3.17 GB
repeat #6
RSS before scan: 3.17 GB
Scan run in 140308511708928
RSS after scan: 3.93 GB
repeat #7
RSS before scan: 3.93 GB
Scan run in 140308503316224
RSS after scan: 4.7 GB
repeat #8
RSS before scan: 4.7 GB
Scan run in 140308494923520
RSS after scan: 5.47 GB
repeat #9
RSS before scan: 5.47 GB
Scan run in 140308486530816
RSS after scan: 6.23 GB
repeat #10
RSS before scan: 6.23 GB
Scan run in 140308478138112
RSS after scan: 7.0 GB
RSS final: 7.0 GB
```

As soon as we reduce the taurus `ThreadPool` size e.g. using this hardcoded change `ThreadPool(Psize=3, daemons=True)` in the script, we reduce the memory footprint:
```
╰─>$ python sardana_scan.py -f /tmp/scan.dat -r 10 -t taurus         
repeat #1
RSS before scan: 105.53 MB
Scan run in 139969951938304
RSS after scan: 106.28 MB
repeat #2
RSS before scan: 106.28 MB
Scan run in 139969943545600
RSS after scan: 722.41 MB
repeat #3
RSS before scan: 722.41 MB
Scan run in 139969935152896
RSS after scan: 1.47 GB
repeat #4
RSS before scan: 1.47 GB
Scan run in 139969951938304
RSS after scan: 1.91 GB
repeat #5
RSS before scan: 1.91 GB
Scan run in 139969943545600
RSS after scan: 2.07 GB
repeat #6
RSS before scan: 2.07 GB
Scan run in 139969935152896
RSS after scan: 2.07 GB
repeat #7
RSS before scan: 2.07 GB
Scan run in 139969951938304
RSS after scan: 2.13 GB
repeat #8
RSS before scan: 2.13 GB
Scan run in 139969943545600
RSS after scan: 2.13 GB
repeat #9
RSS before scan: 2.13 GB
Scan run in 139969935152896
RSS after scan: 2.13 GB
repeat #10
RSS before scan: 2.13 GB
Scan run in 139969951938304
RSS after scan: 2.13 GB
RSS final: 2.13 GB
```

I have also tried a very naive reimplementation of the `RecordList` using the `pandas.DataFrame`, but the behavior does not change:

```
╰─>$ python sardana_scan.py -f /tmp/scan.dat -r 10 -t taurus -p True
repeat #1
RSS before scan: 104.64 MB
Scan run in 139729250084608
RSS after scan: 109.47 MB
repeat #2
RSS before scan: 109.47 MB
Scan run in 139729241691904
RSS after scan: 838.0 MB
repeat #3
RSS before scan: 838.0 MB
Scan run in 139729233299200
RSS after scan: 1.48 GB
repeat #4
RSS before scan: 1.48 GB
Scan run in 139729224906496
RSS after scan: 2.24 GB
repeat #5
RSS before scan: 2.24 GB
Scan run in 139729216513792
RSS after scan: 2.95 GB
repeat #6
RSS before scan: 2.95 GB
Scan run in 139729208121088
RSS after scan: 3.72 GB
repeat #7
RSS before scan: 3.72 GB
Scan run in 139729199728384
RSS after scan: 4.43 GB
repeat #8
RSS before scan: 4.43 GB
Scan run in 139729191335680
RSS after scan: 5.2 GB
repeat #9
RSS before scan: 5.2 GB
Scan run in 139728707909376
RSS after scan: 5.91 GB
repeat #10
RSS before scan: 5.91 GB
Scan run in 139728699516672
RSS after scan: 6.67 GB
RSS final: 6.67 GB
```

I tried using jemalloc2 from Debian Buster but it does not improve the behavior:

```
╰─>$ LD_PRELOAD=/usr/lib/x86_64-linux-gnu/libjemalloc.so.2 python sardana_scan.py -f /tmp/scan.dat -r 10 -t taurus                     
repeat #1
RSS before scan: 112.9 MB
Scan run in 140120511022848
RSS after scan: 973.9 MB
repeat #2
RSS before scan: 973.9 MB
Scan run in 140120497846016
RSS after scan: 1.73 GB
repeat #3
RSS before scan: 1.73 GB
Scan run in 140120483165952
RSS after scan: 2.52 GB
repeat #4
RSS before scan: 2.52 GB
Scan run in 140120468485888
RSS after scan: 3.3 GB
repeat #5
RSS before scan: 3.3 GB
Scan run in 140120453805824
RSS after scan: 4.08 GB
repeat #6
RSS before scan: 4.08 GB
Scan run in 140120439125760
RSS after scan: 4.87 GB
repeat #7
RSS before scan: 4.87 GB
Scan run in 140120424445696
RSS after scan: 5.65 GB
repeat #8
RSS before scan: 5.65 GB
Scan run in 140120409765632
RSS after scan: 6.43 GB
repeat #9
RSS before scan: 6.43 GB
Scan run in 140120395085568
RSS after scan: 7.22 GB
repeat #10
RSS before scan: 7.22 GB
Scan run in 140120380405504
RSS after scan: 7.98 GB
RSS final: 7.98 GB
```

What actually allows to release the memory is to join (shutdown) and garbage collect the threading machinery:

```
╰─>$ python sardana_scan.py -f /tmp/scan.dat -r 10 -t taurus -j True 
repeat #1
RSS before scan: 104.65 MB
Scan run in 139951594694400
RSS after scan: 109.83 MB
repeat #2
RSS before scan: 109.83 MB
Scan run in 139951586301696
RSS after scan: 893.99 MB
repeat #3
RSS before scan: 893.99 MB
Scan run in 139951509337856
RSS after scan: 1.64 GB
repeat #4
RSS before scan: 1.64 GB
Scan run in 139951500945152
RSS after scan: 2.4 GB
repeat #5
RSS before scan: 2.4 GB
Scan run in 139951492552448
RSS after scan: 3.17 GB
repeat #6
RSS before scan: 3.17 GB
Scan run in 139951484159744
RSS after scan: 3.94 GB
repeat #7
RSS before scan: 3.94 GB
Scan run in 139951475767040
RSS after scan: 4.7 GB
repeat #8
RSS before scan: 4.7 GB
Scan run in 139951467374336
RSS after scan: 5.47 GB
repeat #9
RSS before scan: 5.47 GB
Scan run in 139951458981632
RSS after scan: 6.23 GB
repeat #10
RSS before scan: 6.23 GB
Scan run in 139950905358080
RSS after scan: 7.0 GB
RSS final: 1.65 GB
```

Having all this in mind, I see the following options:

1. Recreate a thread on each macro execution (this applies only on the top level macros, internally called macros would still use the same thread). Here the implementation is trivial.
2. Do not store the scan data at all if the `PreserveMacroData=False`. Here the implementation may get complicated in case of the continuous scans using the `ApplyInterpolation=True`

I will measure the time penalty of option 1 and post it soon.

**Comment by reszelaz** [comment]
> 1. Recreate a thread on each macro execution (this applies only on the top level macros, internally called macros would still use the same thread). Here the implementation is trivial.

> I will measure the time penalty of option 1 and post it soon.

I have benchmarked it using [this script](https://github.com/reszelaz/storage-kpi/blob/py3/src/thread_benchmark.py):

1. Creating and joining a thread on each job execution:
```
╰─>$ python thread_benchmark.py -n 100 -e thread                                                                                                                                 
Run 100 times
Run Thread
Mean time: 0.00012078809999366058
```
2. Using taurus `ThreadPool(Psize=10)` to execute jobs (the same as macros are run currently)
```
╰─>$ python thread_benchmark.py -n 100 -e thread_pool
Run 100 times
Run ThreadPool (Psize=10) without join
Mean time: 8.473379999486497e-05
```
3. Creating and joining a `ThreadPool(Psize=1)` on each job execution:
```
╰─>$ python thread_benchmark.py -n 100 -e thread_pool_with_join
Run 100 times
Run ThreadPool (Psize=1) with join
Mean time: 0.00034583077998831866
```
In conclusion, creating a thread for each macro execution would take **0.00012 s** which is a little bit worse than using a thread pool **0.0000847 s**.

Now, I will try to put all these findings together and propose a PR.

**Comment by reszelaz** [comment]
> What still confuses me a bit is when you start a scan the increase at the beginning is very strong and then flattens. Actually, the memory consumption should be the same for each measuring point and therefore the large increase at the beginning doesn't make sense to me.
> Secondly, it is noticeable that at the end of the scan the memory does not return to the initial situation. If the memory is released again after the scan, then the trend should be back to the level before the scan.

Hi @felix92,
In this [article](https://realpython.com/python-memory-management/#cpythons-memory-management) we can read that Python will allocate the memory when a new object is created, and even if the memory is considered free from the Python point of view i.e. object destroyed by the Garbage Collector, the memory may not be returned to the Operating System. This explains why at the beginning of the tests i.e. during the first scan it grows linearly and then it fluctuates.

At the very end of the article, just before the "Conclusion":

> This brings us to the idea of truly freeing memory. You’ll notice that I’ve been saying “free” in quotes quite a bit. The reason is that when a block is deemed “free”, that memory is not actually freed back to the operating system. The Python process keeps it allocated and will use it later for new data. Truly freeing memory returns it to the operating system to use.

> Arenas are the only things that can truly be freed. So, it stands to reason that those arenas that are closer to being empty should be allowed to become empty. That way, that chunk of memory can be truly freed, reducing the overall memory footprint of your Python program.

I still can not find any explanation for worse behavior (more memory allocation) when using multiple _worker threads_.

Cheers!

**Comment by reszelaz** [comment]
After https://github.com/sardana-org/sardana/issues/1095#issuecomment-888928334 
> #1669 that fixes the _mem leak_ with the `NXscanH5_FileRecorder` is already merged.
> Soon #1664 will get merged solving the _mem leak_ with the continuous scans (ascanct, etc.). As soon as this is done I propose to close this issue. The memory of the MacroServer won't grow beyond what is needed for the scan data (there is no progressive accumulation) and IMHO it is not urgent now to look for other solutions here.

@felix92 pointed that the memory was still growing when executing multiple scans but as separate macro executions what appeared to be different with respect to executing them from within the `repeat` macro. This was fixed with #1675.
Meanwhile, #1664 was also merged.

Having all this into account we could say that memory of the MacroServer won't grow beyond what is needed for the scan data (there is no progressive accumulation) and IMHO it is not urgent now to look for other solutions here. So, I propose to close this issue.
